### PR TITLE
Pyfunc get raw model

### LINKED ISSUE #12775: [BUG] Cannot load Scikit models with PyFunc flavor
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### Willingness to contribute

Yes. I can contribute a fix for this bug independently.

### MLflow version

- Client:  2.14.3
- Tracking server:  2.14.3

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: macOS-14.5-arm64-arm-64bit
- **Python version**: 3.9.12
- **yarn version, if running the dev UI**: N/A

### Expected behavior

I should be able to log a model with the `sklearn` flavor, load it back with the `pyfunc` flavor and call `unwrap_python_model()` to retrieve the original model.

### Actual behavior

When calling `unwrap_python_model()` in the described scenario, an exception is raised.

### Tracking information

```shell
System information: Darwin Darwin Kernel Version 23.5.0: Wed May  1 20:12:58 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6000
Python version: 3.9.12
MLflow version: 2.14.3
MLflow module location: /Users/rsayn/Library/Caches/pypoetry/virtualenvs/project-M-gywEHC-py3.9/lib/python3.9/site-packages/mlflow/__init__.py
Tracking URI: file:///Users/rsayn/dev/project/mlruns
Registry URI: file:///Users/rsayn/dev/project/mlruns
MLflow dependencies: 
  Flask: 3.0.3
  Jinja2: 3.1.4
  alembic: 1.13.2
  cachetools: 5.4.0
  click: 8.1.7
  cloudpickle: 3.0.0
  docker: 7.1.0
  entrypoints: 0.4
  gitpython: 3.1.43
  graphene: 3.3
  gunicorn: 22.0.0
  importlib-metadata: 7.2.1
  markdown: 3.6
  matplotlib: 3.9.1
  numpy: 1.26.4
  opentelemetry-api: 1.16.0
  opentelemetry-sdk: 1.16.0
  packaging: 24.1
  pandas: 1.5.3
  protobuf: 4.25.3
  pyarrow: 15.0.2
  pytz: 2024.1
  pyyaml: 6.0.1
  querystring-parser: 1.2.4
  requests: 2.32.3
  scikit-learn: 1.5.1
  scipy: 1.13.1
  sqlalchemy: 2.0.31
  sqlparse: 0.5.1
  virtualenv: 20.26.3
```

### Code to reproduce issue

```python
from sklearn.datasets import load_iris
from sklearn.svm import SVC
import mlflow

X, y = load_iris(return_X_y=True)
svc = SVC().fit(X, y)
with mlflow.start_run() as run:
    mlflow.sklearn.log_model(svc, "model")
    model = mlflow.pyfunc.load_model(f"runs:/{run.info.run_id}/model")
    loaded_svc = model.unwrap_python_model() # here is where it breaks
```

### Stack trace

```python
Traceback (most recent call last):
  File "/Users/rsayn/Library/Caches/pypoetry/virtualenvs/project-M-gywEHC-py3.9/lib/python3.9/site-packages/mlflow/pyfunc/__init__.py", line 872, in unwrap_python_model
    python_model = self._model_impl.python_model
AttributeError: '_SklearnModelWrapper' object has no attribute 'python_model'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/rsayn/dev/project/main_ignore.py", line 13, in <module>
    loaded_svc = model.unwrap_python_model() # here is where it breaks
  File "/Users/rsayn/Library/Caches/pypoetry/virtualenvs/project-M-gywEHC-py3.9/lib/python3.9/site-packages/mlflow/pyfunc/__init__.py", line 876, in unwrap_python_model
    raise MlflowException("Unable to retrieve base model object from pyfunc.") from e
mlflow.exceptions.MlflowException: Unable to retrieve base model object from pyfunc.
```

### Other info / logs

```
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/12814?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12814/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12814
```

</p>
</details>

### Related Issues/PRs

Resolve #12775

### What changes are proposed in this pull request?
Add `get_raw_model` function in pyfunc to get the underlying raw model

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/pyfunc/__init__.py [hunk 1]
```diff
@@ -925,6 +925,15 @@ def __repr__(self):
             info["flavor"] = self._model_meta.flavors[FLAVOR_NAME]["loader_module"]
         return yaml.safe_dump({"mlflow.pyfunc.loaded_model": info}, default_flow_style=False)
 
+    @experimental
+    def get_raw_model(self):
+        """
+        Get the underlying raw model if the model wrapper implemented `get_raw_model` function.
+        """
+        if hasattr(self._model_impl, "get_raw_model"):
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:933]: can we update the wrappers so they all have the `model` attribute?

Comment (serena-ruan) [reply mlflow/pyfunc/__init__.py:933]: Can we instead update all wrappers to have `get_raw_model`? Adding `self.model=...` seems redundant, and we can't directly replace the attribute name either.

Comment (WeichenXu123) [reply mlflow/pyfunc/__init__.py:933]: Q:

Not all models have `get_raw_model` method.
Shall we add  `get_raw_model` for all kinds of models ?

Comment (serena-ruan) [reply mlflow/pyfunc/__init__.py:933]: Yes that's the proposal

File: mlflow/catboost/__init__.py [hunk 2]
```diff
@@ -335,6 +335,12 @@ class _CatboostModelWrapper:
     def __init__(self, cb_model):
         self.cb_model = cb_model
 
+    def get_raw_model(self):
```
Comment (harupy) [inline mlflow/catboost/__init__.py:338]: We might want to add an abstract class. Not a blocker.

Comment (serena-ruan) [reply mlflow/catboost/__init__.py:338]: Yes we really need one! I can file a follow-up PR for this. While `get_raw_model` not always exists, for example `promptlab` it has `model_route` but technically it is not a model.

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**